feat(phoenix-postgres): Inter-app communication (#312)
kraftcloud/examples: phoenix-postgres/README.md

networks:
- uuid:              d14f07de-1c81-4309-bbf3-ab8ab6aeefe6
  private-ip:        10.0.14.53
  mac:               12:b0:0a:00:0e:35
timestamps:
  created:           just now
scale-to-zero:
  enabled:           true
  policy:            on
  cooldown-time:     1s
```

or

**Using the legacy kraft CLI**
```ansi title="kraft"
[●] Deployed successfully!
 │
 ├───────── name: phoenix-hd29m
 ├───────── uuid: fa4db9cf-f7b5-4b8b-b007-8fd1ce87c7e1
 ├──────── metro: https://api.fra.unikraft.cloud/v1
 ├──────── state: starting
 ├─────── domain: https://wild-moon-pkwsqc49.fra.unikraft.app
 ├──────── image: oci://unikraft.io/<my-org>/phoenix@sha256:...
 ├─────── memory: 2048 MiB
 ├────── service: wild-moon-pkwsqc49
 ├─ private fqdn: phoenix-hd29m.internal
 └─── private ip: 10.0.14.53
```

In this case, the instance names are `postgres-ik5at` and `phoenix-hd29m`.
They're different for each run.

Use a browser to access the Phoenix application using the URL from the `fqdn` field in the output.

You can list information about the instances by running:

**Using the unikraft CLI (Recommended)**
```bash title="unikraft"
unikraft instances list
```

```ansi title="unikraft"

> METRO NAME             STATE   IMAGE              MEMORY VCPUS FQDN                                      CREATED

Reviewer: nitpick but there should be 2 spaces between columns

fra   phoenix-hd29m    running <my-org>/phoenix   2GiB   1     wild-moon-pkwsqc49.fra.unikraft.app       just now
fra   postgres-ik5at   standby <my-org>/postgres  2GiB   1     postgres.internal                         3 minutes ago
```

or

**Using the legacy kraft CLI**
```bash title="kraft"
kraft cloud instance list
```

```ansi title="kraft"
NAME             FQDN                                        STATE    STATUS      IMAGE                                              MEMORY    VCPUS  ARGS  BOOT TIME
phoenix-hd29m    wild-moon-pkwsqc49.fra.unikraft.app         running  since now   oci://unikraft.io/<my-org>/phoenix@sha256:...      2048 MiB  1            158.32 ms
postgres-ik5at   postgres.internal                           standby  since 3min  oci://unikraft.io/<my-org>/postgres@sha256:...     2048 MiB  1            1811.99 ms
```

When done, you can remove the instances:

**Using the unikraft CLI (Recommended)**
```bash title="unikraft"
unikraft instances delete postgres-ik5at phoenix-hd29m
```

or

**Using the legacy kraft CLI**
```bash title="kraft"
kraft cloud instance remove postgres-ik5at

> kraft cloud instance remove phoenix-hd29m

Reviewer: ```suggestion
kraft cloud instance remove postgres-ik5at phoenix-hd29m
```

> ## Deployment

Reviewer: ```suggestion
```

To run this example, follow these steps:

1. Install the CLI.
   Use the [unikraft CLI](https://unikraft.com/docs/cli/unikraft) or the legacy [kraft CLI](https://unikraft.org/docs/cli/install).
   You need a [BuildKit](https://github.com/moby/buildkit) builder. The easiest way to get one is via [Docker](https://docs.docker.com/engine/install/).
   Alternatively, you can also directly set up and use BuildKit, see the [quick start](https://github.com/moby/buildkit#quick-start).

> **Note**:
> The unikraft CLI is the current standard, while kraft is the legacy version.

> > Choose one of the CLIs below and only run the commands associated with it for the rest of this guide.

Reviewer: ```suggestion
   > **Note**:
   > The unikraft CLI is the current standard, while kraft is the legacy version.
   > Choose one of the CLIs below and only run the commands associated with it for the rest of this guide.
```

```bash
git clone https://github.com/unikraft-cloud/examples
cd examples/phoenix-postgres/
```

Make sure to log into Unikraft Cloud and pick a [metro](https://unikraft.com/docs/platform/metros) close to you.
This guide uses `fra` (Frankfurt, 🇩🇪):

**Using the unikraft CLI (Recommended)**
```bash title="unikraft"
unikraft login
```

or

**Using the legacy kraft CLI**
```bash title="kraft"
# Set Unikraft Cloud access token
export UKC_TOKEN=token
# Set metro to Frankfurt, DE
export UKC_METRO=fra
```

Create a volume for PostgreSQL data persistence:

**Using the unikraft CLI (Recommended)**
```bash title="unikraft"
unikraft volume create --metro fra --name db-data --size 512M
```

```ansi title="unikraft"
metro:        fra
name:         db-data
uuid:         0efda5f6-21d2-48c7-84f5-8efe87e33090
state:        available
size:         512MiB
filesystem:   ext4
quota-policy: static
persistent:   true
timestamps:
  created:    just now
```

or

**Using the legacy kraft CLI**
```bash title="kraft"
kraft cloud volume create --name db-data --size 512
```

```ansi title="kraft"
[+] Created volume 'db-data' (512 MiB)
```

> Build and deploy the PostgreSQL instance.

Reviewer: add section here ##

```ansi title="unikraft"
metro:                 fra
name:                  postgres-ik5at
uuid:                  3776dbfe-2937-45e7-8079-c54275ef3cff
state:                 starting
image:                 <my-org>/postgres
runtime:
  env:
    PGDATA:            /var/lib/postgresql/data/pgdata
    POSTGRES_DB:       myapp_prod
    POSTGRES_PASSWORD: *
    POSTGRES_USER:     postgres
resources:
  memory:              2GiB
  vcpus:               1
service:
  name:                dry-cloud-q8u7yjkl
  uuid:                cf364e60-46b0-4031-bace-296fa8e24ac4
  domains:
  - fqdn:              postgres.internal
volumes:
- name:                db-data
  uuid:                0efda5f6-21d2-48c7-84f5-8efe87e33090
  at:                  /var/lib/postgresql/data
networks:
- uuid:                cab79d9f-08a8-4741-8efd-58d9a1faa35d
  private-ip:          10.0.14.117
  mac:                 12:b0:0a:00:0e:75
timestamps:
  created:             just now
scale-to-zero:
  enabled:             true
  policy:              idle
  stateful:            true
  cooldown-time:       1s
```

or

**Using the legacy kraft CLI**
```ansi title="kraft"
[●] Deployed successfully!
 │

>  ├───────── name: postgres-o9abw

Reviewer: ideally this would have the same name, uuid etc. as with `unikraft`, since the listing output uses that

 ├───────── uuid: 3776dbfe-2937-45e7-8079-c54275ef3cff
 ├──────── metro: https://api.fra.unikraft.cloud/v1
 ├──────── state: starting
 ├─────── domain: postgres.internal
 ├──────── image: oci://unikraft.io/<my-org>/postgres@sha256:...
 ├─────── memory: 2048 MiB
 ├────── service: dry-cloud-q8u7yjkl
 ├─ private fqdn: postgres-o9abw.internal
 └─── private ip: 10.0.14.117
```

> Build and deploy the Phoenix instance.

Reviewer: add section here ##

Update `SECRET_KEY_BASE` with a unique secret and update the password in `DATABASE_URL` to match the one set for PostgreSQL:

**Using the unikraft CLI (Recommended)**
```bash title="unikraft"
unikraft build ./phoenix --output <my-org>/phoenix:latest

> unikraft run --scale-to-zero policy=on,cooldown-time=1000 --metro fra -p 443:4000/tls+http -m 2G --image <my-org>/phoenix:latest --env SECRET_KEY_BASE=OUT8Y7R7RwaPFp3dZs1pSX2F0A6fra0glidT7VrmKy5e/GJt5zyMjINHYctrl2ZW --env DATABASE_URL=ecto://postgres:example_123@postgres.internal:5432/myapp_prod

Reviewer: valid

```

or

**Using the legacy kraft CLI**
```bash title="kraft"
kraft cloud deploy --scale-to-zero on --scale-to-zero-cooldown 1s -p 443:4000/tls+http -M 2Gi --env SECRET_KEY_BASE=OUT8Y7R7RwaPFp3dZs1pSX2F0A6fra0glidT7VrmKy5e/GJt5zyMjINHYctrl2ZW --env DATABASE_URL=ecto://postgres:example_123@postgres.internal:5432/myapp_prod phoenix/
```

The output shows the instance address and other details:

**Using the unikraft CLI (Recommended)**
```ansi title="unikraft"
metro:               fra
name:                phoenix-hd29m
uuid:                fa4db9cf-f7b5-4b8b-b007-8fd1ce87c7e1
state:               starting
image:               <my-org>/phoenix
runtime:
  env:
    DATABASE_URL:    ecto://postgres:example_123@postgres.internal:5432/myapp_prod

>     SECRET_KEY_BASE: OUT8Y7R7RwaPFp3dZs1pSX2F0A6fra0glidT7VrmKy5e/GJt5zyMjINHYctrl2ZW

Reviewer: just use a placeholder <secret-key-base>

```

In this case, the instance names are `postgres-ik5at` and `phoenix-hd29m`.
They're different for each run.

Use a browser to access the Phoenix application using the URL from the `fqdn` field in the output.

You can list information about the instances by running:

**Using the unikraft CLI (Recommended)**
```bash title="unikraft"
unikraft instances list
```

```ansi title="unikraft"
METRO NAME             STATE   IMAGE              MEMORY VCPUS FQDN                                      CREATED
fra   phoenix-hd29m    running <my-org>/phoenix   2GiB   1     wild-moon-pkwsqc49.fra.unikraft.app       just now
fra   postgres-ik5at   standby <my-org>/postgres  2GiB   1     postgres.internal                          3 minutes ago
```

or

**Using the legacy kraft CLI**
```bash title="kraft"
kraft cloud instance list
```

```ansi title="kraft"
NAME             FQDN                                        STATE    STATUS      IMAGE                                              MEMORY   VCPUS  ARGS  BOOT TIME
phoenix-hd29m    wild-moon-pkwsqc49.fra.unikraft.app          running  since now   oci://unikraft.io/<my-org>/phoenix@sha256:...       2048 MiB  1            158.32 ms

> postgres-ik5at   postgres.internal                           standby  since 3min  oci://unikraft.io/<my-org>/postgres@sha256:...      2048 MiB  1            1811.99 ms

Reviewer: pls align

## Deployment

To run this example, follow these steps:

1. Install the CLI.
   Use the [unikraft CLI](https://unikraft.com/docs/cli/unikraft) or the legacy [kraft CLI](https://unikraft.org/docs/cli/install).
   You need a [BuildKit](https://github.com/moby/buildkit) builder. The easiest way to get one is via [Docker](https://docs.docker.com/engine/install/).
   Alternatively, you can also directly set up and use BuildKit, see the [quick start](https://github.com/moby/buildkit#quick-start).

2. Clone the [`examples` repository](https://github.com/unikraft-cloud/examples) and `cd` into the `examples/phoenix-postgres` directory:

```bash
git clone https://github.com/unikraft-cloud/examples
cd examples/phoenix-postgres/
```

Make sure to log into Unikraft Cloud and pick a [metro](https://unikraft.com/docs/platform/metros) close to you.
This guide uses `fra` (Frankfurt, 🇩🇪):

```bash title="unikraft"
unikraft login
```

or

```bash title="kraft"
# Set Unikraft Cloud access token
export UKC_TOKEN=token
# Set metro to Frankfurt, DE
export UKC_METRO=fra
```

> 3. Create a volume for PostgreSQL data persistence:

Reviewer: Remove list numbering from here onwards

```bash title="unikraft"
unikraft volume create --metro fra --name db-data --size 512M
```

or

```bash title="kraft"
kraft cloud volume create --name db-data --size 512Mi
```

4. Build and deploy the PostgreSQL instance.
   Update `POSTGRES_PASSWORD` with a secure password:

```bash title="unikraft"

> unikraft build postgres/ --output <my-org>/postgres:latest

Reviewer: ```suggestion
unikraft build ./postgres --output <my-org>/postgres:latest
```

> unikraft run --scale-to-zero policy=idle,cooldown-time=1000,stateful=true --metro fra -m 2G --volume db-data:/var/lib/postgresql/data --image <my-org>/postgres:latest --domain=postgres.internal --env POSTGRES_USER=postgres --env POSTGRES_PASSWORD=example_123 --env POSTGRES_DB=myapp_prod --env PGDATA=/var/lib/postgresql/data/pgdata

Reviewer: ```suggestion
unikraft run --scale-to-zero policy=idle,cooldown-time=1000,stateful=true --metro fra -m 2G --volume db-data:/var/lib/postgresql/data --image <my-org>/postgres:latest --domain postgres.internal --env POSTGRES_USER=postgres --env POSTGRES_PASSWORD=example_123 --env POSTGRES_DB=myapp_prod --env PGDATA=/var/lib/postgresql/data/pgdata
```

```

or

```bash title="kraft"

> kraft cloud deploy -M 2G --scale-to-zero idle --scale-to-zero-cooldown 1s --scale-to-zero-stateful --volume db-data:/var/lib/postgresql/data --domain=postgres.internal postgres/ --env POSTGRES_USER=postgres --env POSTGRES_PASSWORD=example_123 --env POSTGRES_DB=myapp_prod --env PGDATA=/var/lib/postgresql/data/pgdata

Reviewer: ```suggestion
kraft cloud deploy -M 2Gi --scale-to-zero idle --scale-to-zero-cooldown 1s --scale-to-zero-stateful --volume db-data:/var/lib/postgresql/data --domain postgres.internal ./postgres --env POSTGRES_USER=postgres --env POSTGRES_PASSWORD=example_123 --env POSTGRES_DB=myapp_prod --env PGDATA=/var/lib/postgresql/data/pgdata
```

```

The output shows the PostgreSQL instance details:

```ansi title="kraft"
[●] Deployed successfully!
 │
 ├───────── name: postgres-vuxpy
 ├───────── uuid: 27da06cd-a7ec-4127-9da4-a06b0884e54c
 ├──────── metro: https://api.fra.unikraft.cloud/v1
 ├──────── state: starting
 ├─────── domain: postgres.internal
 ├──────── image: oci://unikraft.io/<my-org>/postgres@sha256:9844d19a2e69c5e4fc95ee05695bf19d5c2580360382cdf37dcae921fb6440ca

>  ├─────── memory: 1907 MiB

Reviewer: ```suggestion
 ├─────── memory: 2048 MiB
```

 ├────── service: damp-orangutan-690nn6ln
 ├─ private fqdn: postgres-vuxpy.internal
 └─── private ip: 10.0.0.37
```

or

```ansi title="unikraft"
metro:                 fra

> name:                  postgres-tlwkd

Reviewer: ```suggestion
name:                  postgres-vuxpy
```

> uuid:                  afa55279-b3aa-4d20-8e74-400f57cf3a0a

Reviewer: ```suggestion
uuid:                  27da06cd-a7ec-4127-9da4-a06b0884e54c
```

state:                 starting
image:                 <my-org>/postgres
runtime:
  env:
    PGDATA:            /var/lib/postgresql/data/pgdata
    POSTGRES_DB:       myapp_prod
    POSTGRES_PASSWORD: *
    POSTGRES_USER:     postgres
resources:
  memory:              2GiB
  vcpus:               1
service:
  uuid:                a7d8c9e9-3732-4bd6-a39a-95c735ad4346

>   name:                broken-paper-fg599stf

Reviewer: ```suggestion
  name:                damp-orangutan-690nn6ln
```

  domains:
  - fqdn:              postgres.internal
volumes:
- db-data:/var/lib/postgresql/data
networks:
- uuid:                3faffbd0-9169-443b-9471-e5af47a32c5a
  private-ip:          10.0.0.37
  mac:                 12:b0:0a:00:00:25
timestamps:
  created:             just now
scale-to-zero:         policy=idle,stateful=true,cooldown-time=1s
```

5. Build and deploy the Phoenix instance.
   Update `SECRET_KEY_BASE` with a unique secret and update the password in `DATABASE_URL` to match the one set for PostgreSQL:

```bash title="unikraft"

> unikraft build myapp/ --output <my-org>/myapp:latest

Reviewer: ```suggestion
unikraft build ./phoenix --output <my-org>/phoenix:latest
```

> unikraft run --scale-to-zero policy=idle,cooldown-time=1000,stateful=true --metro fra -p 443:4000/tls+http -m 2G --image <my-org>/myapp:latest --env SECRET_KEY_BASE=OUT8Y7R7RwaPFp3dZs1pSX2F0A6fra0glidT7VrmKy5e/GJt5zyMjINHYctrl2ZW --env DATABASE_URL=ecto://postgres:example_123@postgres.internal:5432/myapp_prod

Reviewer: ```suggestion
unikraft run --scale-to-zero policy=idle,cooldown-time=1000,stateful=true --metro fra -p 443:4000/tls+http -m 2G --image <my-org>/phoenix:latest --env SECRET_KEY_BASE=OUT8Y7R7RwaPFp3dZs1pSX2F0A6fra0glidT7VrmKy5e/GJt5zyMjINHYctrl2ZW --env DATABASE_URL=ecto://postgres:example_123@postgres.internal:5432/myapp_prod
```

```

or

```bash title="kraft"

> kraft cloud deploy -M 2G --scale-to-zero idle --scale-to-zero-cooldown 1s --scale-to-zero-stateful -p 443:4000/tls+http myapp/ --env SECRET_KEY_BASE=OUT8Y7R7RwaPFp3dZs1pSX2F0A6fra0glidT7VrmKy5e/GJt5zyMjINHYctrl2ZW --env DATABASE_URL=ecto://postgres:example_123@postgres.internal:5432/myapp_prod

Reviewer: ```suggestion
kraft cloud deploy -M 2Gi --scale-to-zero idle --scale-to-zero-cooldown 1s --scale-to-zero-stateful -p 443:4000/tls+http ./phoenix --env SECRET_KEY_BASE=OUT8Y7R7RwaPFp3dZs1pSX2F0A6fra0glidT7VrmKy5e/GJt5zyMjINHYctrl2ZW --env DATABASE_URL=ecto://postgres:example_123@postgres.internal:5432/myapp_prod
```

```

The output shows the instance address and other details:

```ansi title="kraft"
[●] Deployed successfully!
 │
 ├───────── name: phoenix-t5620
 ├───────── uuid: 142dfa6c-9448-4ede-ad7e-80e899e41d66
 ├──────── metro: https://api.fra.unikraft.cloud/v1
 ├──────── state: starting
 ├─────── domain: https://broken-frost-yidthcen.fra.unikraft.app
 ├──────── image: oci://unikraft.io/<my-org>/phoenix@sha256:adbec137ec92f697a41bbd6e95b24611b927b0e430d632e080a12bac57b67f17

>  ├─────── memory: 1907 MiB

Reviewer: ```suggestion
 ├─────── memory: 2048 MiB
```

 ├────── service: broken-frost-yidthcen
 ├─ private fqdn: phoenix-t5620.internal
 └─── private ip: 10.0.0.17
```

or

```ansi title="unikraft"
metro:               fra

> name:                myapp-9ad2u

Reviewer: ```suggestion
name:                phoenix-t5620
```

> uuid:                fe1ada23-bae8-4cc5-aad8-d833b1cd31db

Reviewer: ```suggestion
uuid:                142dfa6c-9448-4ede-ad7e-80e899e41d66
```

state:               starting

> image:               <my-org>/myapp

Reviewer: ```suggestion
image:               <my-org>/phoenix-t5620
```

runtime:
  env:
    DATABASE_URL:    ecto://postgres:example_123@postgres.internal:5432/myapp_prod
    SECRET_KEY_BASE: OUT8Y7R7RwaPFp3dZs1pSX2F0A6fra0glidT7VrmKy5e/GJt5zyMjINHYctrl2ZW
resources:
  memory:            2GiB
  vcpus:             1
service:
  uuid:              2e9cc96e-0e7c-432f-b9e8-60fa9de1bb3d

>   name:              still-resonance-5s85egfk

Reviewer: ```suggestion
  name:              broken-frost-yidthcen
```

  domains:

>   - fqdn:            still-resonance-5s85egfk.fra.unikraft.app

Reviewer: ```suggestion
  - fqdn:            broken-frost-yidthcen.fra.unikraft.app
```

networks:
- uuid:              f353051b-0f34-4c74-a19d-66b2f357a15c
  private-ip:        10.0.0.17
  mac:               12:b0:0a:00:00:11
timestamps:
  created:           just now
scale-to-zero:       policy=idle,stateful=true,cooldown-time=1s
```

> In this case, the instance names are `postgres-tlwkd` and `myapp-9ad2u`.

Reviewer: ```suggestion
In this case, the instance names are `postgres-vuxpy` and `phoenix-t5620`.
```

They're different for each run.

Use a browser to access the Phoenix application using the URL from the `domain` / `fqdn` field in the output.

> 

Reviewer: Add listing output for the instances:
`You can list information about the instances by running: ... `

When done, you can remove the instances:

```bash title="unikraft"
unikraft instances delete postgres-tlwkd

> unikraft instances delete myapp-9ad2u

Reviewer: ```suggestion
unikraft instances delete postgres-tlwkd phoenix-t5620
```

```

or

```bash title="kraft"
kraft cloud instance remove postgres-vuxpy

> kraft cloud instance remove phoenix-t5620

Reviewer: ```suggestion
kraft cloud instance remove postgres-vuxpy phoenix-t5620
```
